### PR TITLE
Register Operations logic

### PR DESCRIPTION
**Pull request by andreiblt1304**

### CODE REVIEW

File: multisigverifier/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+        if is_bls_valid && signatures_count > 2/3 * bls_pub_keys {
+            return true
+        }
+
+        false
+    }
+
+    #[storage_mapper("isValid")]
+    fn is_valid(&self) -> SingleValueMapper<bool>;
```
Comment (dorin-iancu) [inline multisigverifier/src/lib.rs:61]: Unused?

File: multisigverifier/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    #[storage_mapper("bls_pub_keys")]
+    fn bls_pub_keys(&self) -> VecMapper<ManagedAddress>;
+
+    #[storage_mapper("signers")]
+    fn signatures(&self) -> SingleValueMapper<u32>;
```
Comment (dorin-iancu) [inline multisigverifier/src/lib.rs:67]: Mismatch of names. Use `signaturesCount` for mapper name and `signature_count` for function name.

File: multisigverifier/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    #[storage_mapper("operations_mapper")]
+    fn pending_operations_mapper(&self) -> SetMapper<ManagedBuffer>;
```
Comment (dorin-iancu) [inline multisigverifier/src/lib.rs:70]: Use `UnorderedSetMapper` instead, way more efficient.

File: multisigverifier/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        let is_bls_valid = self.crypto().verify_bls(
+            user.as_managed_buffer(), 
+            &serialized_signature_data, 
+            signature.as_managed_buffer()
+        );
+
+        let signatures_count = self.signatures().get();
+        let bls_pub_keys = self.bls_pub_keys().len() as u32;
+
+        if is_bls_valid && signatures_count > 2/3 * bls_pub_keys {
```
Comment (dorin-iancu) [inline multisigverifier/src/lib.rs:53]: This won't work. `2/3` is 0. Change to `> 2 * bls_pub_key / 3`, and maybe even have a separate function for it, like `get_min_signatures`

File: multisigverifier/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    fn register_bridge_operations(
+        &self,
+        hash_of_hashes: ManagedBuffer,
+        hash_of_bridge_ops: MultiValueEncoded<ManagedBuffer>,
+        signature: BlsSignature<Self::Api>,
+    ) {
+        let caller = self.blockchain().get_caller();
+        let is_bls_valid = self.verify_bls(hash_of_bridge_ops.clone(), &signature, caller);
+        
+        if is_bls_valid {
```
Comment (dorin-iancu) [inline multisigverifier/src/lib.rs:26]: Change this to a `require!` instead.

File: multisigverifier/src/lib.rs [hunk 6]
```diff
@@ -16,29 +19,45 @@ pub trait Multisigverifier:
...
+        
+        require!(
+            bridge_operations_hash.eq(transfers_data_hash),
+            "Hash of all operations doesn't match the hash of transfer data"
+        );
+
+        require!(
+            is_bls_valid,
+            "BLS signature is not valid"
+        );
```
Comment (dorin-iancu) [inline multisigverifier/src/lib.rs:46]: Do this check right after the `verify_bls` call, no need to process further if that fails.

File: multisigverifier/src/lib.rs [hunk 7]
```diff
@@ -47,25 +66,27 @@ pub trait Multisigverifier:
...
         let signatures_count = self.signatures().get();
         let bls_pub_keys = self.bls_pub_keys().len() as u32;
+        let minimum_signatures = 2 * bls_pub_keys / 3;
 
-        if is_bls_valid && signatures_count > 2/3 * bls_pub_keys {
+        if is_bls_valid && signatures_count > minimum_signatures {
             return true
         }
 
         false
```
Comment (dorin-iancu) [inline multisigverifier/src/lib.rs:81]: Can simply return the condition. i.e.

is_bls_valid && signatures_count > minimum_signatures

Having said that, I wouldn't give a bool arg to this function just to check if it's true, I'd just check it outside the function.

Comment (andreiblt1304) [reply multisigverifier/src/lib.rs]: done

File: multisigverifier/src/lib.rs (hunk 6 above)
Comment (andreiblt1304) [reply multisigverifier/src/lib.rs]: done

File: multisigverifier/tests/multisigverifier_test.rs
Comment (CostinCarabas) [inline multisigverifier/tests/multisigverifier_test.rs:1]: Remove empty file

File: multisigverifier/tests/multisig_balckbox_test.rs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+#[test]
+fn test_register_bridge_ops() {
+    let mut state = MultisigTestState::new();
+    state.deploy_multisig_contract();
+
+    let bridge_operations_hash = "6ee1e00813a74f8293d2c63172c062d38bf780d8811ff63984813a49cd61ff9e";
+    let mock_signature: BlsSignature<StaticApi> = ManagedByteArray::new_from_bytes(
+        b"EIZ2\x05\xf7q\xc7G\x96\x1f\xba0\xe2\xd1\xf5pE\x14\xd7?\xac\xff\x8d\x1a\x0c\x11\x900f5\xfb\xff4\x94\xb8@\xc5^\xc2,exn0\xe3\xf0\n"
```
Comment (CostinCarabas) [inline multisigverifier/tests/multisig_balckbox_test.rs:94]: How did you get this magic signature? Would be useful to compute it here.

Comment (andreiblt1304) [reply multisigverifier/tests/multisig_balckbox_test.rs:94]: I tried using methods from the SDK but couldn't find the correct steps

File: multisigverifier/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,106 @@
+#![no_std]
+
+use core::ptr::null;
```
Comment (dorin-iancu) [inline multisigverifier/src/lib.rs:3]: clippy: unused import

Comment (andreiblt1304) [reply multisigverifier/src/lib.rs:3]: resolving more clippy warning right now

**Review by dorin-iancu — APPROVED** [review]

**Pull request merged**